tests/main/cgroup-tracking/task.yaml

```diff
@@ -0,0 +1,104 @@
+summary: Each snap app and hook is tracked via cgroups
+# Ubuntu 14.04's special version of systemd doesn't have StartTransientUnit API.
+systems: [-ubuntu-14.04-*]
+details: |
+    This test creates a snap process that suspends itself and ensures that it
+    placed into the appropriate hierarchy.
+prepare: |
+    # This feature depends on the release-app-awareness feature
+    snap set core experimental.refresh-app-awareness=true
+
+    #shellcheck source=tests/lib/snaps.sh
+    . "$TESTSLIB"/snaps.sh
+    install_local test-snapd-sh
+
+restore: |
+    snap unset core experimental.refresh-app-awareness
+
+execute: |
+    # TODO: add a variant that runs as an user
+    # TODO: add a variant that uses a service
+
+    # This test varies between Ubuntu 16.04, Ubuntu 18.04 and Fedora 31.
+    # This combination exercises each of the three cases below.
+    echo "Find the path and id of the cgroup snapd uses for tracking."
+    if grep -q -F "0::" < /proc/self/cgroup; then
+        # Hybrid or pure v2 mode available.
+        if [ "$(stat -f --print=%T /sys/fs/cgroup)" = "cgroup2fs" ]; then
+            base_cg_path=/sys/fs/cgroup
+        else
+            base_cg_path=/sys/fs/cgroup/unified
+        fi
+        base_cg_id=0
+    else
+        # Only name=systemd available
+        base_cg_path=/sys/fs/cgroup/systemd
+        base_cg_id="$(grep -F 'name=systemd' < /proc/self/cgroup | cut -d : -f 1)"
+    fi
+
+    echo "Sanity check, base directory of selected cgroup exists."
+    test -d "$base_cg_path"
+
+    echo "Start a \"sleep\" process in the background"
+    #shellcheck disable=SC2016
+    test-snapd-sh.sh -c 'touch $SNAP_DATA/1.stamp && exec sleep 3m' &
+    pid1=$!
+    trap 'killall sleep || true' EXIT
+    echo "Ensure that snap-confine has finished its task and that the snap process"
+    echo "is active. Note that we don't want to wait forever either."
+    retry-tool -n 30 --wait 0.1 test -e /var/snap/test-snapd-sh/current/1.stamp
+
+    echo "During startup snap-confine has asked systemd to move the process to a"
+    echo "new transient scope. The scope name is \"snap.\$random.test-snapd-sh.sh\"."
+    echo "Let's verify that."
+    pid1_tracking_cg_path="$(grep -E "^$base_cg_id:" < "/proc/$pid1/cgroup" | cut -d : -f 3)"
+    echo "$pid1_tracking_cg_path" | MATCH '.*/snap\.[0-9a-f-]+\.test-snapd-sh\.sh\.scope'
```

Comment:
I tried this myself by building your branch locally, installing it on bionic, running snap set core experimental.refresh-app-awareness=true and then using test-snapd-sh.sh.

How I built it:
```
$ git clone --single-branch --branch feature/refresh-app-awareness-v2 https://github.com/zyga/snapd.git snapd.zyga
$ cd snapd.zyga
$ schroot -c bionic-amd64 -u root
$ sudo apt-get build-dep snapd
$ sudo apt-get install golang-1.10
$ dpkg-buildpackage -rfakeroot  # creates ../snapd_2.43.2_amd64.deb
```

Eg, in one terminal:
```
jamie@sec-bionic-amd64:~$ sudo dpkg -i ./snapd_2.43.2_amd64.deb  # from the above build
jamie@sec-bionic-amd64:~$ snap version
snap    2.43.2+git23025.ge9f4770
snapd   2.43.2+git23025.ge9f4770
series  16
ubuntu  18.04
kernel  4.15.0-76-generic

jamie@sec-bionic-amd64:~$ sudo snap set core experimental.refresh-app-awareness=true
jamie@sec-bionic-amd64:~$ sudo snap get core experimental.refresh-app-awareness
true
jamie@sec-bionic-amd64:~$ sudo snap install test-snapd-sh
jamie@sec-bionic-amd64:~$ test-snapd-sh.sh
$ set |grep SNAP_NAME
SNAP_NAME='test-snapd-sh'
```

Then in another terminal I did:
```
jamie@sec-bionic-amd64:~$ ps aZ|grep test-snapd
snap.test-snapd-sh.sh (enforce)  2036 pts/0    S+     0:00 /bin/sh

jamie@sec-bionic-amd64:~$ cat /proc/2036/cgroup
12:freezer:/snap.test-snapd-sh
11:devices:/user.slice
10:hugetlb:/
9:cpuset:/
8:cpu,cpuacct:/
7:pids:/user.slice/user-1000.slice/session-4.scope
6:perf_event:/
5:rdma:/
4:blkio:/
3:memory:/
2:net_cls,net_prio:/
1:name=systemd:/user.slice/user-1000.slice/session-4.scope
0::/user.slice/user-1000.slice/session-4.scope
```

Note, I do not see the "snap.$random.test-snapd-sh.sh"....

Reply:
I haven't reproduced this yet but are you, by any chance, falling the victim of re-execution? Looking at the version numbers you may be using edge rather than the local build.

Having said that I'll double check. We have a spread test so I'm pretty sure this really works but I would like to give you an educated explanation if I can.

Reply:
I checked snap version before installing the deb and the version (from stable) was less than what snap version showed after installing the deb. I can try again after disabling reexec explicitly.

Reply:
I built snapd from source and got this to work on the first try: 

This is from my Ubuntu 20.04 system.
```
zyga      256492  7.7  1.4 1675352 116188 pts/0  Sl+  19:00   0:00 /snap/spotify/41/usr/share/spotify/spotify
zyga      256602  0.1  0.7 556440 61616 pts/0    S+   19:00   0:00 /snap/spotify/41/usr/share/spotify/spotify --type=zygote --no-sandbox --log-file=/snap/spotify/41/usr/share/spotify/debug.
zyga      256627  0.7  0.9 936988 79336 pts/0    Sl+  19:00   0:00 /snap/spotify/41/usr/share/spotify/spotify --type=utility --field-trial-handle=16092028785744898111,5012669361373076538,13
zyga      256632  3.7  1.2 1409224 103632 pts/0  Sl+  19:00   0:00 /snap/spotify/41/usr/share/spotify/spotify --type=renderer --no-sandbox --log-file=/snap/spotify/41/usr/share/spotify/debu
zyga      256649  1.1  1.2 885224 97600 pts/0    Sl+  19:00   0:00 /snap/spotify/41/usr/share/spotify/spotify --type=gpu-process --field-trial-handle=16092028785744898111,501266936137307653
zyga      256665  0.1  0.0  22156  5512 pts/1    Ss   19:00   0:00 bash
zyga      256673  0.0  0.0  22700  3404 pts/1    R+   19:00   0:00 ps aux
zyga@focal:~/go/src/github.com/snapcore$ cat /proc/256492/cgroup 
12:rdma:/
11:cpuset:/
10:memory:/user.slice/user-1000.slice/user@1000.service
9:devices:/snap.spotify.spotify
8:net_cls,net_prio:/
7:freezer:/snap.spotify
6:hugetlb:/
5:perf_event:/
4:pids:/user.slice/user-1000.slice/user@1000.service
3:cpu,cpuacct:/
2:blkio:/
1:name=systemd:/user.slice/user-1000.slice/user@1000.service/snap.bceecc40-6503-4740-af85-6dd4bbaad4f5.spotify.spotify.scope
0::/user.slice/user-1000.slice/user@1000.service/snap.bceecc40-6503-4740-af85-6dd4bbaad4f5.spotify.spotify.scope
```

I need to inspect bionic next.

Reply:
Ok, I tried again with with all the fixed debugging commits and changes from today and I was able to make the work for root, but not non-root.

### as non-root
In one terminal:
```
$ SNAPD_DEBUG=1 test-snapd-sh.sh
2020/02/18 14:22:10.497486 cmd_linux.go:177: DEBUG: re-exec disabled by user
2020/02/18 14:22:10.507006 cmd_run.go:365: DEBUG: SELinux not enabled
2020/02/18 14:22:10.507994 cmd_run.go:1007: DEBUG: creating transient scope snap.test-snapd-sh.sh
DEBUG: umask reset, old umask was   02
DEBUG: security tag: snap.test-snapd-sh.sh
DEBUG: executable:   /usr/lib/snapd/snap-exec
DEBUG: confinement:  non-classic
DEBUG: base snap:    core
DEBUG: ruid: 1000, euid: 0, suid: 0
DEBUG: rgid: 1000, egid: 0, sgid: 0
DEBUG: apparmor label on snap-confine is: /usr/lib/snapd/snap-confine
DEBUG: apparmor mode is: enforce
DEBUG: creating lock directory /run/snapd/lock (if missing)
DEBUG: opening lock directory /run/snapd/lock
DEBUG: opening lock file: /run/snapd/lock/.lock
DEBUG: sanity timeout initialized and set for 30 seconds
DEBUG: acquiring exclusive lock (scope (global), uid 0)
DEBUG: sanity timeout reset and disabled
DEBUG: ensuring that snap mount directory is shared
DEBUG: unsharing snap namespace directory
DEBUG: releasing lock 5
DEBUG: opened snap-update-ns executable as file descriptor 5
DEBUG: opened snap-discard-ns executable as file descriptor 6
DEBUG: creating lock directory /run/snapd/lock (if missing)
DEBUG: opening lock directory /run/snapd/lock
DEBUG: opening lock file: /run/snapd/lock/test-snapd-sh.lock
DEBUG: sanity timeout initialized and set for 30 seconds
DEBUG: acquiring exclusive lock (scope test-snapd-sh, uid 0)
DEBUG: sanity timeout reset and disabled
DEBUG: initializing mount namespace: test-snapd-sh
DEBUG: snappy_udev_init
DEBUG: forked support process 3467
DEBUG: block device of snap core, revision 8039 is 7:0
DEBUG: sanity timeout initialized and set for 30 seconds
DEBUG: joining preserved mount namespace for inspection
DEBUG: DEBUG: block device of the root filesystem is 7:0
changing apparmor hat to mount-namespace-capture-helperDEBUG: sanity timeout reset and disabled

DEBUG: DEBUG: preserved mount is not stale, reusing
DEBUG: joined preserved mount namespace test-snapd-sh
DEBUG: joining preserved per-user mount namespace
DEBUG: unsharing the mount namespace (per-user)
DEBUG: sc_setup_user_mounts: test-snapd-sh
DEBUG: NOT preserving per-user mount namespace
DEBUG: moved process 3455 to cgroup hierarchy /sys/fs/cgroup/freezer/snap.test-snapd-sh
DEBUG: releasing lock 7
DEBUG: sending command 0 to helper process (pid: 3467)
DEBUG: waiting for response from helper
helper process waiting for command
DEBUG: sanity timeout initialized and set for 30 seconds
DEBUG: sanity timeout reset and disabled
DEBUG: helper process received command 0
DEBUG: waiting for the helper process to exit
DEBUG: helper process exiting
DEBUG: helper process exited normally
DEBUG: resetting PATH to values in sync with core snap
DEBUG: creating user data directory: /home/jamie/snap/test-snapd-sh/2
DEBUG: requesting changing of apparmor profile on next exec to snap.test-snapd-sh.sh
DEBUG: ruid: 1000, euid: 1000, suid: 0
DEBUG: setting capabilities bounding set
DEBUG: regaining SYS_ADMIN
DEBUG: loading bpf program for security tag snap.test-snapd-sh.sh
DEBUG: read 6320 bytes from /var/lib/snapd/seccomp/bpf//snap.test-snapd-sh.sh.bin
DEBUG: read 152 bytes from /var/lib/snapd/seccomp/bpf/global.bin
DEBUG: clearing SYS_ADMIN
DEBUG: execv(/usr/lib/snapd/snap-exec, /usr/lib/snapd/snap-exec...)
DEBUG:  argv[1] = test-snapd-sh.sh
DEBUG: umask restored to   02
DEBUG: working directory restored to /home/jamie
$
```

In another terminal:
```
$ find /sys/fs/cgroup -name "*scope*"|grep snap
$

$ ps aZ|grep snapd
...
snap.test-snapd-sh.sh (enforce)  3599 pts/0    S+     0:00 /bin/sh
...

$ cat /proc/3599/cgroup 
12:perf_event:/
11:freezer:/snap.test-snapd-sh
10:pids:/user.slice/user-1000.slice/session-4.scope
9:devices:/user.slice
8:hugetlb:/
7:blkio:/
6:memory:/
5:cpuset:/
4:rdma:/
3:cpu,cpuacct:/
2:net_cls,net_prio:/
1:name=systemd:/user.slice/user-1000.slice/session-4.scope
0::/user.slice/user-1000.slice/session-4.scope
```

So it does not work as non-root.

### as root
In one terminal:
```
$ sudo SNAPD_DEBUG=1 test-snapd-sh.sh
2020/02/18 14:29:34.897672 cmd_linux.go:177: DEBUG: re-exec disabled by user
2020/02/18 14:29:34.903706 cmd_run.go:365: DEBUG: SELinux not enabled
2020/02/18 14:29:34.904036 cmd_run.go:1007: DEBUG: creating transient scope snap.test-snapd-sh.sh
DEBUG: umask reset, old umask was  022
DEBUG: security tag: snap.test-snapd-sh.sh
DEBUG: executable:   /usr/lib/snapd/snap-exec
DEBUG: confinement:  non-classic
DEBUG: base snap:    core
DEBUG: ruid: 0, euid: 0, suid: 0
DEBUG: rgid: 0, egid: 0, sgid: 0
DEBUG: apparmor label on snap-confine is: /usr/lib/snapd/snap-confine
DEBUG: apparmor mode is: enforce
DEBUG: creating lock directory /run/snapd/lock (if missing)
DEBUG: opening lock directory /run/snapd/lock
DEBUG: opening lock file: /run/snapd/lock/.lock
DEBUG: sanity timeout initialized and set for 30 seconds
DEBUG: acquiring exclusive lock (scope (global), uid 0)
DEBUG: sanity timeout reset and disabled
DEBUG: ensuring that snap mount directory is shared
DEBUG: unsharing snap namespace directory
DEBUG: releasing lock 5
DEBUG: opened snap-update-ns executable as file descriptor 5
DEBUG: opened snap-discard-ns executable as file descriptor 6
DEBUG: creating lock directory /run/snapd/lock (if missing)
DEBUG: opening lock directory /run/snapd/lock
DEBUG: opening lock file: /run/snapd/lock/test-snapd-sh.lock
DEBUG: sanity timeout initialized and set for 30 seconds
DEBUG: acquiring exclusive lock (scope test-snapd-sh, uid 0)
DEBUG: sanity timeout reset and disabled
DEBUG: initializing mount namespace: test-snapd-sh
DEBUG: snappy_udev_init
DEBUG: forked support process 3575
DEBUG: block device of snap core, revision 8039 is 7:0
DEBUG: sanity timeout initialized and set for 30 seconds
DEBUG: joining preserved mount namespace for inspection
DEBUG: changing apparmor hat to mount-namespace-capture-helperDEBUG: 
DEBUG: helper process waiting for command
DEBUG: sanity timeout initialized and set for 30 seconds
block device of the root filesystem is 7:0
DEBUG: sanity timeout reset and disabled
DEBUG: preserved mount is not stale, reusing
DEBUG: joined preserved mount namespace test-snapd-sh
DEBUG: moved process 3563 to cgroup hierarchy /sys/fs/cgroup/freezer/snap.test-snapd-sh
DEBUG: releasing lock 7
DEBUG: sending command 0 to helper process (pid: 3575)
DEBUG: waiting for response from helper
DEBUG: sanity timeout reset and disabled
DEBUG: helper process received command 0
DEBUG: waiting for the helper process to exit
DEBUG: helper process exiting
DEBUG: helper process exited normally
DEBUG: resetting PATH to values in sync with core snap
DEBUG: creating user data directory: /root/snap/test-snapd-sh/2
DEBUG: requesting changing of apparmor profile on next exec to snap.test-snapd-sh.sh
DEBUG: ruid: 0, euid: 0, suid: 0
DEBUG: loading bpf program for security tag snap.test-snapd-sh.sh
DEBUG: read 6320 bytes from /var/lib/snapd/seccomp/bpf//snap.test-snapd-sh.sh.bin
DEBUG: read 152 bytes from /var/lib/snapd/seccomp/bpf/global.bin
DEBUG: execv(/usr/lib/snapd/snap-exec, /usr/lib/snapd/snap-exec...)
DEBUG:  argv[1] = test-snapd-sh.sh
DEBUG: umask restored to  022
DEBUG: working directory restored to /home/jamie
#
```

In another terminal:
```
$ find /sys/fs/cgroup -name "*scope*"|grep snap
/sys/fs/cgroup/pids/system.slice/snap.022c73ea-bfb7-4a12-a984-75a2a774625a.test-snapd-sh.sh.scope
/sys/fs/cgroup/devices/system.slice/snap.022c73ea-bfb7-4a12-a984-75a2a774625a.test-snapd-sh.sh.scope
/sys/fs/cgroup/systemd/system.slice/snap.022c73ea-bfb7-4a12-a984-75a2a774625a.test-snapd-sh.sh.scope
/sys/fs/cgroup/unified/system.slice/snap.022c73ea-bfb7-4a12-a984-75a2a774625a.test-snapd-sh.sh.scope

$ ps aZ|grep snapd
...
snap.test-snapd-sh.sh (enforce)  3563 pts/0    S+     0:00 /bin/sh
...

$ cat /proc/3563/cgroup 
12:perf_event:/
11:freezer:/snap.test-snapd-sh
10:pids:/system.slice/snap.022c73ea-bfb7-4a12-a984-75a2a774625a.test-snapd-sh.sh.scope
9:devices:/system.slice/snap.022c73ea-bfb7-4a12-a984-75a2a774625a.test-snapd-sh.sh.scope
8:hugetlb:/
7:blkio:/
6:memory:/
5:cpuset:/
4:rdma:/
3:cpu,cpuacct:/
2:net_cls,net_prio:/
1:name=systemd:/system.slice/snap.022c73ea-bfb7-4a12-a984-75a2a774625a.test-snapd-sh.sh.scope
0::/system.slice/snap.022c73ea-bfb7-4a12-a984-75a2a774625a.test-snapd-sh.sh.scope
```

So, it works as root, but not non-root.

Reply:
> 9:devices:/system.slice/snap.022c73ea-bfb7-4a12-a984-75a2a774625a.test-snapd-sh.sh.scope

This got me wondering what is going to happen with the device cgroup that snap-confine sets up. So I checked:

In one terminal:
```
$ sudo snap install chocolate-doom-jdstrand # eg, it plugs joystick so gets the device cgroup
$ sudo snap run --shell chocolate-doom-jdstrand.chocolate-doom
#
```

In another terminal:
```
$ ps aZ|grep choc
...
snap.chocolate-doom-jdstrand.chocolate-doom (enforce) 4012 pts/0 S+   0:00 /bin/bash
...

$ cat /proc/4012/cgroup 
...
9:devices:/snap.chocolate-doom-jdstrand.chocolate-doom
...
```

Which looks correct. This probably underscores an opportunity to work better within systemd's cgroup handling, but that is not needed for this PR as snap-confine continues to move the process out of systemd's device cgroup into the snap's per-snap-command cgroup, as designed.

Reply:
I've expanded the test that runs as an user and am investigating what is going on.

Reply:
I've expanded the test that runs as an user and am investigating what is going on.

Reply:
```
google:ubuntu-18.04-64 .../tests/main/cgroup-tracking# su -l test -c 'test-snapd-sh.sh -c "cat /proc/self/cgroup"'
internal error, please report: running "test-snapd-sh.sh" failed: cannot create transient scope: Process org.freedesktop.systemd1 exited with status 1
```

Corresponding journal:
```
Feb 19 10:42:26 feb191031-578298 su[19650]: Successful su for test by root
Feb 19 10:42:26 feb191031-578298 su[19650]: + /dev/pts/0 root:test
Feb 19 10:42:26 feb191031-578298 su[19650]: pam_unix(su:session): session opened for user test by root(uid=0)
Feb 19 10:42:26 feb191031-578298 su[19650]: pam_systemd(su:session): Cannot create session: Already running in a session
Feb 19 10:42:26 feb191031-578298 dbus-daemon[19677]: [session uid=12345 pid=19675] AppArmor D-Bus mediation is enabled
Feb 19 10:42:26 feb191031-578298 dbus-daemon[19677]: [session uid=12345 pid=19675] Activating service name='org.freedesktop.systemd1' requested by ':1.0' (uid=12345 pid=19651 comm="test-sna
Feb 19 10:42:26 feb191031-578298 dbus-daemon[19677]: [session uid=12345 pid=19675] Activated service 'org.freedesktop.systemd1' failed: Process org.freedesktop.systemd1 exited with status 1
Feb 19 10:42:26 feb191031-578298 su[19650]: pam_unix(su:session): session closed for user test
```

Reply:
This is even more interesting:
```
google:ubuntu-18.04-64 .../tests/main/cgroup-tracking# su test -c 'snap run test-snapd-sh.sh -c "cat /proc/self/cgroup"'
error: dial unix /run/user/0/bus: connect: permission denied
```

I think our idea of running as user, in tests, using su, is broken.

Reply:
I'm debugging this and I have some observations

Using an 18.04 desktop system I can use `systemd-run --user --scope` correctly so we are not missing anything on this particular front.

The way the test is handling user switching (with my yet-to-be-published) changes is incorrect and insufficient for using DBus.

I cannot explain the failure you saw yet.

Reply:
Do note for your spread tests and debugging, in my python dbus probing of the issue, on bionic, in the console/ssh scenario, there was also a wrinkle that the transient scope seemed partially started in that I would get UnitExists errors when I ran my python script a second time and I couldn't clean this up with StopUnit or KillUnit (there was nothing running to kill or stop).

I *think* so long as you are killing your systemd user session correctly you should be ok for spread, but fyi for testing.

Reply:
After some additional discussion on IRC I now understand the scenario in which this failed for Jamie:

- a user logging into a bionic system remotely over ssh
- a user logging into a bionic system interactively over vt

I'll debug those cases and try to expand tests to cover them, if possible.

Reply:
We now have the session tool and we can implement those tests. This should uncover the failure you saw and allow us to get to the bottom of the problem.

Reply:
I'll further add that adding the transient scope worked perfectly on 20.04 via:
- desktop session as uid 1000
- ssh session as uid 1000
- console session as uid 1000
- desktop session as uid 1001
- ssh session as uid 1000
- console session as uid 1000
- ssh session calling snap under sudo
- ssh session calling snap under sudo -u <user with uid 1001>

(note, uid 1000 is in all the standard 'first user' groups but 1001 is not and has no supplementary groups).